### ISSUE TITLE
Getting slower in contiguous tests with quic proxy

### ISSUE THREAD

**Issue by ByungjinJun** [lucas-clemente/quic-go#1586]
Hi,

I have been working on creating QUIC tunnel to provide a faster connection for users of high loss / high latency network. My proxy implemented your quic-go and works great. Thank you, first of all.

However, I have an issue on testing with headless chrome. My proxy serves TCP and QUIC at the same time, but only when I test with QUIC tunnel, the test get slower after around 150~180 tests are done without any problem. This never happens when I test with TCP tunnel. For your understanding, I would like to provide the system structure of my proxy in the simplest way.

client(10.10.10.10)
client proxy <--QUIC or TCP tunnel--> server proxy(10.10.20.20) -> target server

Basically TCP and QUIC proxy use the same structure of the code, except dialer in the client proxy and listener in the server proxy. TCP just uses a built-in net library and QUIC uses quic-go. I think other parts of codes are ok since TCP works completely fine. Also, I don't think my server causes the problem since it always responds to the client as soon as it gets a request, and it starts working fine again when I restart the client side.
So I suspect the problem is caused by my QUIC client, specifically my QUIC dialer. The core part of the QUIC Dialer code is like below. Dialer simply uses DialAddr in quic-go.

**QUIC dialer**
```
// QUICDialer creates a QUIC Dialer with QuicConfig
func QUICDialer(config *QuicConfig) Dialer {
	if config == nil {
		config = &QuicConfig{}
	}

	if config.TLSConfig == nil {
		config.TLSConfig = &tls.Config{InsecureSkipVerify: true}
	}

	return &quicDialer{
		config: config,
		sessions: make(map[string]quic.Session),
	}
}
```

```
func (dr *quicDialer) Dial(addr string) (net.Conn, error) {
	// set quic-go config for DialAddr
	qConfig := &quic.Config{
		HandshakeTimeout: QUICHandshakeTimeout,
		IdleTimeout:      QUICIdleTimeout,
		KeepAlive:        dr.config.KeepAlive,
	}

	dr.Mutex.Lock()
	defer dr.Mutex.Unlock()

	session, ok := dr.sessions[addr]
	// Initially quicDialer has no quic session
	if !ok || session == nil {
		// quic.DialAddr Resolves UDP addr and Listen
		s, err := quic.DialAddr(addr, dr.config.TLSConfig, qConfig)
		if err != nil {
			log.Println("[QUIC CL ERR] dial session failed:", err)
		}
		session = s
		dr.sessions[addr] = session
	}

	stream, err := session.OpenStreamSync()
        // remove the error handling part <- this never happened in my testing

        // net.Conn
	return &quicConn{
		Stream:     stream,
		localAddr:  session.LocalAddr(),
		remoteAddr: session.RemoteAddr(),
	}, nil
}
```

Is there anything missing or I am doing wrong? Please help me out if you have any idea.

**Comment by marten-seemann** [comment]
Thanks for raising this issue. It's interesting that this only occurs after 150 tests or so, which might hint to a congestion related problem, or a resource exhaustion.
Have you checked the memory and CPU consumption over time? Maybe we're leaking memory or go routines somewhere.

**Comment by marten-seemann** [comment]
This should probably have gotten a lot better with the recent fixes to our pacing algorithm (#1855 and #1857). Feel free to open a new issue if the problem still exists.
